mllib/src/main/scala/org/apache/spark/ml/feature/OneHotEncoderEstimator.scala

```diff
@@ -0,0 +1,519 @@
+    val existingFields = schema.fields
+
+    require(inputColNames.length == outputColNames.length,
+      s"The number of input columns ${inputColNames.length} must be the same as the number of " +
+        s"output columns ${outputColNames.length}.")
+
+    // Input columns must be NumericType.
+    inputColNames.foreach(SchemaUtils.checkNumericType(schema, _))
+
+    // Prepares output columns with proper attributes by examining input columns.
+    val inputFields = $(inputCols).map(schema(_))
+
+    val outputFields = inputFields.zip(outputColNames).map { case (inputField, outputColName) =>
+      OneHotEncoderCommon.transformOutputColumnSchema(
+        inputField, outputColName, dropLast, keepInvalid)
+    }
+    outputFields.foldLeft(schema) { case (newSchema, outputField) =>
+      SchemaUtils.appendColumn(newSchema, outputField)
+    }
+  }
+}
+
+/**
+ * A one-hot encoder that maps a column of category indices to a column of binary vectors, with
+ * at most a single one-value per row that indicates the input category index.
+ * For example with 5 categories, an input value of 2.0 would map to an output vector of
+ * `[0.0, 0.0, 1.0, 0.0]`.
+ * The last category is not included by default (configurable via `dropLast`),
+ * because it makes the vector entries sum up to one, and hence linearly dependent.
+ * So an input value of 4.0 maps to `[0.0, 0.0, 0.0, 0.0]`.
+ *
+ * @note This is different from scikit-learn's OneHotEncoder, which keeps all categories.
+ * The output vectors are sparse.
+ *
+ * When `handleInvalid` is configured to 'keep', an extra "category" indicating invalid values is
+ * added as last category. So when `dropLast` is true, invalid values are encoded as all-zeros
+ * vector.
+ *
+ * @note When encoding multi-column by using `inputCols` and `outputCols` params, input/output cols
+ * come in pairs, specified by the order in the arrays, and each pair is treated independently.
+ *
+ * @see `StringIndexer` for converting categorical values into category indices
+ */
+@Since("2.3.0")
+class OneHotEncoderEstimator @Since("2.3.0") (@Since("2.3.0") override val uid: String)
+    extends Estimator[OneHotEncoderModel] with OneHotEncoderBase with DefaultParamsWritable {
+
+  @Since("2.3.0")
+  def this() = this(Identifiable.randomUID("oneHotEncoder"))
+
+  /** @group setParam */
+  @Since("2.3.0")
+  def setInputCols(values: Array[String]): this.type = set(inputCols, values)
+
+  /** @group setParam */
+  @Since("2.3.0")
+  def setOutputCols(values: Array[String]): this.type = set(outputCols, values)
+
+  /** @group setParam */
+  @Since("2.3.0")
+  def setDropLast(value: Boolean): this.type = set(dropLast, value)
+
+  /** @group setParam */
+  @Since("2.3.0")
+  def setHandleInvalid(value: String): this.type = set(handleInvalid, value)
+
+  @Since("2.3.0")
+  override def transformSchema(schema: StructType): StructType = {
+    // When fitting data, we want the the plain number of categories without `handleInvalid` and
```

Comment:
This is for fitting to get the plain number of categories based on the input schema. Here we just get the plain numbers from the input schema and record them into the trained model.

In model transform, transformSchema takes them into account.

Reply:
Estimator.transformSchema should return an accurate schema for whatever Estimator.fit().transform() would produce.  That's what the API requires.  The schema this method is returning right now is not accurate since it does not take the Params handleInvalid and dropLast into account.

I agree the Model should store the plain number of categories, but that's a separate issue from what transformSchema is supposed to do.

Reply:
During fitting, we obtain the plain number of categories from the metadata in the schema first, and then from input data if no metadata available.

If `transformSchema` has the requirement to return accurate output schema even for `Estimator` (because I think estimator doesn't output data but model), then I think I need to convert the number of categories with handleInvalid and dropLast taken into account back to plain number before storing into the model.

Reply:
Why do you need to change the numCategories which the model stores?

Model.transformSchema already has logic for taking the raw numCategories and computing a schema which takes the Params into account.  Can that not be reused for Estimator.transformSchema?

I just think it's a lot simpler to store the raw numCategories in the Model.  (It would be really confusing for those values to be based off of 1 setting of Params, and then for the user to change the Params, and for us to therefore need to store both original and new values for Params.)

Reply:
There is misunderstanding. The model stores the raw numCategories now. I don't want to change it.

For now, the transformSchema logic shared between Model and Estimator is the same: it uses the metadata from input schema to compute the output schema with the Params into account or not.

The numCategories is determined from this schema then. As you require Estimator.transformSchema to return a schema with the Params into account, the numCategories derived from the schema is not raw now. But we want to record raw numCategories into the model, so I need to compute raw numCategories from the numCategories derived the schema.

Another approach is, I let Estimator.transformSchema to return a schema with the Params into account. When I need to record the raw numCategories, I call validateAndTransformSchema again to get the raw numCategories without the Param into account.

Reply:
Oh, sorry, I did misunderstand.  I like your last suggestion:

> Another approach is, I let Estimator.transformSchema to return a schema with the Params into account. When I need to record the raw numCategories into the model, I call validateAndTransformSchema again to get the raw numCategories without the Param into account.

This will be nice for reusing code + give the desired output for transformSchema.

Thank you!